crates/iota-json-rpc/src/governance_api.rs

```diff
@@ -642,36 +655,71 @@ fn pending_and_candidate_validators_exchange_rate(
                 false,
             )
         })
-        .collect::<Vec<(IotaAddress, ObjectID, ObjectID, u64, bool)>>();
+        .collect::<Vec<ValidatorTable>>();
+
+    validator_exchange_rates(state, tables)
+}
+
+/// Check for validators in the `Candidate` state and get its exchange rate. For
+/// these validators, their exchange rates should not be cached as their state
+/// can occur during an epoch or across multiple ones. In contrast, exchange
+/// rates for `Active` and `Inactive` validators can be cached, as their state
+/// changes only at epoch change.
+fn candidate_validators_exchange_rate(
+    state: &Arc<dyn StateRead>,
+) -> RpcInterimResult<Vec<ValidatorExchangeRates>> {
+    let system_state_summary = state.get_system_state()?.into_iota_system_state_summary();
 
-    // Try to find for validator candidate and get its exchange rates
-    for (_, df) in state.get_dynamic_fields(
+    // From validator_candidates_id table get validator info using as key its
+    // IotaAddress
+    let tables = validator_summary_from_system_state(
+        state,
         system_state_summary.validator_candidates_id,
-        None,
-        system_state_summary.validator_candidates_size as usize,
-    )? {
-        let raw_address =
-            String::deserialize(df.name.value).map_err(|err| IotaError::from(err.to_string()))?;
-        let validator_address = IotaAddress::from_str(&raw_address)?;
-
-        let validator_summary = get_validator_from_table(
-            object_store,
-            system_state_summary.validator_candidates_id,
-            &validator_address,
-        )?;
-
-        tables.push((
-            validator_summary.iota_address,
-            validator_summary.staking_pool_id,
-            validator_summary.exchange_rates_id,
-            validator_summary.exchange_rates_size,
-            false,
-        ));
-    }
+        system_state_summary.validator_candidates_size,
+        |df| {
+            bcs::from_bytes::<IotaAddress>(&df.bcs_name).map_err(|err| {
+                IotaError::ObjectDeserialization {
+                    error: err.to_string(),
+                }
+                .into()
+            })
+        },
+    )?;
 
     validator_exchange_rates(state, tables)
 }
 
+/// An utility function to gather `Candidate` or `Inactive` validator status
+/// information from the system state
```

Comment:
I think this could use some refinement. When I read it I don't understand how  `Candidate` or `Inactive` are relevant to the function's signature.

I would suggest first discussing the closure `key` (what it does, how it is used) which seems like more generic and then mention `Candidate` or `Inactive` being fetched with realization of `key`. An example could be useful as well.

Reply:
[Fixed](https://github.com/iotaledger/iota/pull/4313/commits/2547ea52607206b7ae132cfd626c8b5783fe12da#diff-0a10b5dd25f3cf15826ae9c8e8ab585ac0e76d35d7c0f84a9334f7ed722b5406R678-R723)